File: php/ProductsSample.php
<?php
/**
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once 'BaseSample.php';

// Class for running through some example interactions with the
// Products service.
class ProductsSample extends BaseSample {
  // These constants define the identifiers for all of our example products
  // The products will be sold online
  const CHANNEL = 'online';
  // The product details are provided in English
  const CONTENT_LANGUAGE = 'en';
  // The products are sold in the United States
  const TARGET_COUNTRY = 'US';

  // This constant defines how many example products to create in a batch
  const BATCH_SIZE = 10;

  public function run() {
    if (is_null($this->session->websiteUrl)) {
      throw InvalidArgumentException(
          'Cannot run Products workflow on a Merchant Center account without '
          . 'a configured website URL.');
    }
    $exampleProductId = 'book123';
    $exampleProduct = $this->createExampleProduct($exampleProductId);

    $this->insertProduct($exampleProduct);

    // There is a short period after creating a product during which it may not
    // be retrieved (get). In general use it would be unusual to do so anyway,
    // but for the purpose of this example we retry with back off.
    $this->session->retry($this, 'getProduct', $exampleProductId, 10);
    $this->updateProduct($exampleProduct);
    $this->deleteProduct($exampleProductId);

    $exampleProductBatchIDs = [];

    for ($i = 0; $i < self::BATCH_SIZE; $i++) {
      $exampleProductBatchIDs[] = 'book' . $i;
    }

    $exampleProductBatch =
      $this->createExampleProducts($exampleProductBatchIDs);
    $this->insertProductBatch($exampleProductBatch);
    $this->listProducts();
    $this->deleteProductBatch($exampleProductBatchIDs);
  }

  public function insertProduct(
      Google_Service_ShoppingContent_Product $product) {
    $response = $this->session->service->products->insert(
        $this->session->merchantId, $product);
    printf ("Product created: %s \n", $response->getId());

    // Our example product generator does not set a product_type
    // We will do it within an update call.
  }

  public function getProduct($offerId) {
    $productId = $this->buildProductId($offerId);
    $product = $this->session->service->products->get(
        $this->session->merchantId, $productId);
    printf("Retrieved product %s: %s \n", $product->getId(),
        $product->getTitle());
  }

  public function updateProduct(
      Google_Service_ShoppingContent_Product $product) {
    // Let's add a product_type and update the product
    $product->setProductTypes(['English/Classics']);
    // Notice that we use insert. The products service does not have an update
    // method. Inserting a product with an ID that already exists means the same
    // as doing an update anyway.
    $response = $this->session->service->products->insert(
        $this->session->merchantId, $product);
    printf ("Product updated: %s \n", $response->getId());

  }

  public function deleteProduct($offerId) {
    $productId = $this->buildProductId($offerId);
    // The response for a successful delete is empty
    $this->session->service->products->delete(
        $this->session->merchantId, $productId);
    printf ("Product deleted: %s \n", $offerId);
  }

  public function insertProductBatch($products) {
    $entries = [];

    foreach ($products as $key => $product) {
      $entry =
          new Google_Service_ShoppingContent_ProductsCustomBatchRequestEntry();
      $entry->setMethod('insert');
      $entry->setBatchId($key);
      $entry->setProduct($product);
      $entry->setMerchantId($this->session->merchantId);

      $entries[] = $entry;
    }

    $batchRequest =
        new Google_Service_ShoppingContent_ProductsCustomBatchRequest();
    $batchRequest->setEntries($entries);

    $batchResponse =
        $this->session->service->products->custombatch($batchRequest);

    printf("Inserted %d products.\n", count($batchResponse->entries));

    foreach ($batchResponse->entries as $entry) {
      if (empty($entry->getErrors())) {
        $product = $entry->getProduct();
      } else {
        print ("There were errors inserting a product:\n");
        foreach ($entry->getErrors()->getErrors() as $error) {
          printf("\t%s\n", $error->getMessage());
        }
      }
    }
  }

  public function listProducts() {
    // We set the maximum number of results to be lower than the number of
    // products that we inserted, to demonstrate paging.
    $parameters = ['maxResults' => self::BATCH_SIZE - 1];
    $products = $this->session->service->products->listProducts(
        $this->session->merchantId, $parameters);
    $count = 0;
    // You can fetch all items in a loop. We limit to looping just 3
    // times for this example as it may take a long time to finish if you
    // already have many products.
    while (!empty($products->getResources()) && $count++ < 3) {
      foreach ($products->getResources() as $product) {
        printf("%s %s\n", $product->getId(), $product->getTitle());
      }
      // If the result has a nextPageToken property then there are more pages
      // available to fetch
      if (empty($products->getNextPageToken())) {
        break;
      }
      // You can fetch the next page of results by setting the pageToken
      // parameter with the value of nextPageToken from the previous result.
      $parameters['pageToken'] = $products->nextPageToken;
      $products = $this->session->service->products->listProducts(
          $this->session->merchantId, $parameters);
    }
  }

  public function deleteProductBatch($offerIds) {
    $entries = [];

    foreach ($offerIds as $key => $offerId) {
      $entry =
          new Google_Service_ShoppingContent_ProductsCustomBatchRequestEntry();
      $entry->setMethod('delete');
      $entry->setBatchId($key);
      $entry->setProductId($this->buildProductId($offerId));
      $entry->setMerchantId($this->session->merchantId);

      $entries[] = $entry;
    }

    $batchRequest =
        new Google_Service_ShoppingContent_ProductsCustomBatchRequest();
    $batchRequest->setEntries($entries);

    $batchResponses =
        $this->session->service->products->custombatch($batchRequest);
    $errors = 0;
    foreach ($batchResponses->entries as $entry) {
      if (!empty($entry->getErrors())) {
        $errors++;
      }
    }
    print "Requested delete of batch inserted test products\n";
    printf("There were %d errors\n", $errors);
  }

  private function buildProductId($offerId) {
    return sprintf('%s:%s:%s:%s', self::CHANNEL, self::CONTENT_LANGUAGE,
      self::TARGET_COUNTRY, $offerId);
  }

  private function createExampleProducts($offerIds) {
    $products = [];

    foreach ($offerIds as $offerId) {
      $products[] = $this->createExampleProduct($offerId);
    }

    return $products;
  }

  private function createExampleProduct($offerId) {
    $product = new Google_Service_ShoppingContent_Product();

    $product->setOfferId($offerId);
    $product->setTitle('A Tale of Two Cities');
    $product->setDescription('A classic novel about the French Revolution');
    $product->setLink($this->session->websiteUrl . '/tale-of-two-cities.html');
    $product->setImageLink(
        $this->session->websiteUrl . '/tale-of-two-cities.jpg');
    $product->setContentLanguage(self::CONTENT_LANGUAGE);
    $product->setTargetCountry(self::TARGET_COUNTRY);
    $product->setChannel(self::CHANNEL);
    $product->setAvailability('in stock');
    $product->setCondition('new');
    $product->setGoogleProductCategory('Media > Books');
    $product->setGtin('9780007350896');

    $price = new Google_Service_ShoppingContent_Price();
    $price->setValue('2.50');
    $price->setCurrency('USD');

    $product->setPrice($price);

    $shippingPrice = new Google_Service_ShoppingContent_Price();
    $shippingPrice->setValue('0.99');
    $shippingPrice->setCurrency('USD');

    $shipping = new Google_Service_ShoppingContent_ProductShipping();
    $shipping->setPrice($shippingPrice);
    $shipping->setCountry('US');
    $shipping->setService('Standard shipping');

    $product->setShipping([$shipping]);

    $shippingWeight =
        new Google_Service_ShoppingContent_ProductShippingWeight();
    $shippingWeight->setValue(200);
    $shippingWeight->setUnit('grams');

    $product->setShippingWeight($shippingWeight);

    return $product;
  }
}

File: dotnet/source/ProductsSample.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using Google.Apis.ShoppingContent.v2_1;
using Google.Apis.ShoppingContent.v2_1.Data;

namespace ShoppingSamples.Content
{
    /// <summary>
    /// A sample consumer that runs multiple requests against the Products
    /// service in the Content API for Shopping.
    /// <para>These include:
    /// <list type="bullet">
    /// <item>
    /// <description>Products.insert</description>
    /// </item>
    /// <item>
    /// <description>Products.custombatch</description>
    /// </item>
    /// <item>
    /// <description>Products.get</description>
    /// </item>
    /// <item>
    /// <description>Products.list</description>
    /// </item>
    /// <item>
    /// <description>Products.delete</description>
    /// </item>
    /// </list></para>
    /// </summary>
    public class ProductsSample
    {
        private ShoppingContentService service;
        private int maxListPageSize;
        ShoppingUtil shoppingUtil = new ShoppingUtil();
        IEnumerable<HttpStatusCode> retryCodes = new HttpStatusCode[] {HttpStatusCode.NotFound};

        /// <summary>Initializes a new instance of the <see cref="ProductsSample"/> class.</summary>
        /// <param name="service">Content service object on which to run the requests.</param>
        /// <param name="maxListPageSize">The maximum page size to retrieve.</param>
        public ProductsSample(ShoppingContentService service, int maxListPageSize)
        {
            this.service = service;
            this.maxListPageSize = maxListPageSize;
        }

        /// <summary>Runs multiple requests against the Products service.</summary>
        internal void RunCalls(ulong merchantId, string websiteUrl = null)
        {
            if (websiteUrl == null)
            {
                Console.WriteLine("Cannot run Products workflow without a configured website URL.");
                return;
            }
            // Product insertion
            Product newProduct = InsertProduct(merchantId, websiteUrl);
            List<String> productList = InsertProductCustombatch(merchantId, websiteUrl);

            GetAllProducts(merchantId);

            UpdateProduct(merchantId, newProduct.Id);

            // To show the inserted (single) product has changed.
            GetAllProducts(merchantId);

            DeleteProduct(merchantId, newProduct);
            DeleteProductCustombatch(merchantId, productList);
        }

        /// <summary>Gets and prints all products for the given merchant ID.</summary>
        /// <returns>The last page of retrieved accounts.</returns>
        private ProductsListResponse GetAllProducts(ulong merchantId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Listing all Products");
            Console.WriteLine("=================================================================");

            // Retrieve account list in pages and display data as we receive it.
            string pageToken = null;
            ProductsListResponse productsResponse = null;

            do
            {
                ProductsResource.ListRequest accountRequest = service.Products.List(merchantId);
                accountRequest.MaxResults = maxListPageSize;
                accountRequest.PageToken = pageToken;

                productsResponse = accountRequest.Execute();

                if (productsResponse.Resources != null && productsResponse.Resources.Count != 0)
                {
                    foreach (var product in productsResponse.Resources)
                    {
                        Console.WriteLine(
                            "Product with ID \"{0}\" and title \"{1}\" was found.",
                            product.Id,
                            product.Title);
                    }
                }
                else {
                    Console.WriteLine("No accounts found.");
                }

                pageToken = productsResponse.NextPageToken;
            } while (pageToken != null);
            Console.WriteLine();

            // Return the last page of accounts.
            return productsResponse;
        }

        /// <summary>
        /// Updates a product using the products.insert method as there is no update method.
        /// <para>Inserting a product with an ID that already exists means the same as doing an update.</para>
        /// </summary>
        private void UpdateProduct(ulong merchantId, String productId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine(String.Format("Updating product {0}", productId));
            Console.WriteLine("=================================================================");
            // First we need to retrieve the full object, since there are no partial updates for the
            // products collection in Content API v2.1.

            var request = service.Products.Get(merchantId, productId);
            Product product = shoppingUtil.ExecuteWithRetries(request, retryCodes);

            // Set ETag to null as Insert() will reject it otherwise.
            product.ETag = null;

            product.ProductTypes = new List<String>();
            product.ProductTypes.Add("English/Classics");

            // Before inserting, product.Source needs to be cleared.
            product.Source = null;

            Product response = service.Products.Insert(product, merchantId).Execute();
            Console.WriteLine(
                "Product updated with ID \"{0}\" and title \"{1}\".",
                response.Id,
                response.Title);
            Console.WriteLine();
        }

        /// <summary>
        /// Adds a product to the specified account.
        /// </summary>
        /// <returns>The inserted product</returns>
        private Product InsertProduct(ulong merchantId, string websiteUrl)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Inserting a product");
            Console.WriteLine("=================================================================");
            Product product = GenerateProduct(websiteUrl);

            Product response = service.Products.Insert(product, merchantId).Execute();
            Console.WriteLine(
                "Product inserted with ID \"{0}\" and title \"{1}\".",
                response.Id,
                response.Title);
            Console.WriteLine();
            return response;
        }

        /// <summary>
        /// Deletes a product from the specified account.
        /// </summary>
        private void DeleteProduct(ulong merchantId, Product product)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine(String.Format("Deleting product {0}", product.Id));
            Console.WriteLine("=================================================================");

            service.Products.Delete(merchantId, product.Id).Execute();

            Console.WriteLine(
                "Product with ID \"{0}\" and title \"{1}\" was deleted.",
                product.Id,
                product.Title);
            Console.WriteLine();
        }

        /// <summary>
        /// Inserts several products to the specified account, using custombatch.
        /// </summary>
        /// <returns>The list of product IDs, which can be used to get or delete them.</returns>
        private List<String> InsertProductCustombatch(ulong merchantId, string websiteUrl)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Inserting products using custombatch");
            Console.WriteLine("=================================================================");

            ProductsCustomBatchRequest batchRequest = new ProductsCustomBatchRequest();
            batchRequest.Entries = new List<ProductsCustomBatchRequestEntry>();
            for (int i = 0; i < 3; i++)
            {
                ProductsCustomBatchRequestEntry entry = new ProductsCustomBatchRequestEntry();
                entry.BatchId = i;
                entry.MerchantId = merchantId;
                entry.Method = "insert";
                entry.Product = GenerateProduct(websiteUrl);
                batchRequest.Entries.Add(entry);
            }

            ProductsCustomBatchResponse response = service.Products.Custombatch(batchRequest).Execute();
            List<String> productsInserted = new List<String>();
            if (response.Kind == "content#productsCustomBatchResponse")
            {
                for (int i = 0; i < response.Entries.Count; i++)
                {
                    Product product = response.Entries[i].Product;
                    productsInserted.Add(product.Id);
                    Console.WriteLine(
                        "Product inserted with ID \"{0}\" and title \"{1}\".",
                        product.OfferId,
                        product.Title);
                }
            }
            else {
                Console.WriteLine(
                    "There was an error. Response: {0}",
                    response.ToString());
            }
            return productsInserted;

        }

        /// <summary>
        /// Deletes several products from the specified account, using custombatch.
        /// </summary>
        private void DeleteProductCustombatch(ulong merchantId, List<String> productList)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Deleting products using custombatch");
            Console.WriteLine("=================================================================");

            ProductsCustomBatchRequest batchRequest = new ProductsCustomBatchRequest();
            batchRequest.Entries = new List<ProductsCustomBatchRequestEntry>();
            for (int i = 0; i < productList.Count; i++)
            {
                ProductsCustomBatchRequestEntry entry = new ProductsCustomBatchRequestEntry();
                entry.BatchId = i;
                entry.MerchantId = merchantId;
                entry.Method = "delete";
                entry.ProductId = productList[i]; // Use the full product ID here, not the OfferId
                batchRequest.Entries.Add(entry);
            }

            ProductsCustomBatchResponse response = service.Products.Custombatch(batchRequest).Execute();
            if (response.Kind == "content#productsCustomBatchResponse")
            {
                for (int i = 0; i < response.Entries.Count; i++)
                {
                    Errors errors = response.Entries[i].Errors;
                    if (errors != null)
                    {
                        for (int j = 0; j < errors.ErrorsValue.Count; j++)
                        {
                            Console.WriteLine(errors.ErrorsValue[j].ToString());
                        }
                    }
                    else {
                        Console.WriteLine("Product deleted, batchId {0}", response.Entries[i].BatchId);
                    }
                }
            }
            else {
                Console.WriteLine(
                    "There was an error. Response: {0}",
                    response);
            }
        }

        internal Product GenerateProduct(string websiteUrl)
        {
              Product product = new Product() {
              OfferId = String.Format("product#{0}", shoppingUtil.GetUniqueId()),
              Title = "A Tale of Two Cities",
              Description = "A classic novel about the French Revolution",
              Link = $"{websiteUrl}/tale-of-two-cities.html",
              ImageLink = $"{websiteUrl}/tale-of-two-cities.jpg",
              ContentLanguage = "EN",
              TargetCountry = "US",
              Channel = "online",
              Availability = "in stock",
              Condition = "new",
              GoogleProductCategory = "Media > Books",
              Gtin = "9780007350896",
              Price = new Price() {
                Currency = "USD",
                Value = "2.50"
              },
              Shipping = new List<ProductShipping> {
                new ProductShipping() {
                  Country = "US",
                  Service = "Standard shipping",
                  Price = new Price() {
                    Currency = "USD",
                    Value = "0.99"
                  }
                }
              },
              ShippingWeight = new ProductShippingWeight() {
                Unit = "grams",
                Value = 200
              }
            };

            return product;
        }
    }
}
